src/csrs/service/organisationalUnitService.ts

```diff
@@ -1,51 +1,137 @@
+import { getLogger } from '../../utils/logger';
+import { OrganisationalUnitClient } from '../client/organisationalUnitClient';
+import { AgencyToken } from '../model/agencyToken';
+import { OrganisationalUnit } from '../model/organisationalUnit';
+import { OrganisationalUnitPageModel } from '../model/organisationalUnitPageModel';
+import { OrganisationalUnitTypeAhead } from '../model/organisationalUnitTypeAhead';
+import { OrganisationalUnitCache } from '../organisationalUnitCache';
+import { OrganisationalUnitTypeaheadCache } from '../organisationalUnitTypeaheadCache';
 
 export class OrganisationalUnitService {
 	logger = getLogger('OrganisationalUnitService')
-	csrs: Csrs
-	organisationalUnitFactory: OrganisationalUnitFactory
-	agencyTokenCapacityUsedHttpService: AgencyTokenCapacityUsedHttpService
 
-	constructor(csrs: Csrs, organisationalUnitFactory: OrganisationalUnitFactory, agencyTokenCapacityUsedHttpService: AgencyTokenCapacityUsedHttpService) {
-		this.csrs = csrs
-		this.organisationalUnitFactory = organisationalUnitFactory
-		this.agencyTokenCapacityUsedHttpService = agencyTokenCapacityUsedHttpService
+	constructor(private readonly organisationalUnitCache: OrganisationalUnitCache,
+		private readonly organisationalUnitTypeaheadCache : OrganisationalUnitTypeaheadCache,
+		private readonly organisationalUnitClient: OrganisationalUnitClient,
+		private readonly agencyTokenCapacityUsedService: AgencyTokenCapacityUsedHttpService) { }
+
+	async getOrgDropdown(): Promise<OrganisationalUnitTypeAhead> {
+		let typeahead = await this.organisationalUnitTypeaheadCache.getTypeahead()
+		if (typeahead === undefined) {
+			typeahead = await this.refreshTypeahead()
+		}
+		return typeahead
 	}
 
-	async getOrganisationalUnit(uri: string) {
-		let organisationalUnit: any
-		let parent: any
+	async getOrgHierarchy(organisationId: number, hierarchy: OrganisationalUnit[] = []): Promise<OrganisationalUnit[]> {
+        let org = await this.organisationalUnitCache.get(organisationId)
+		if (org == null) {
+			let orgWithAllParents = await this.getOrganisationFromApi(organisationId, true)
+			return orgWithAllParents.getHierarchyAsArray()
+		} else {
+			hierarchy.push(org)
+			if (org.parentId) {
+				return await this.getOrgHierarchy(org.parentId, hierarchy)
+			}
+		}
+		return hierarchy
+    }
 
-		organisationalUnit = await this.csrs.getOrganisationalUnit(uri).catch(error => {
-			throw error
-		})
+	async getOrgTree(): Promise<OrganisationalUnit[]> {
+		const dropdown = await this.getOrgDropdown()
+		return dropdown.getAsTree()
+	}
 
-		parent = await this.csrs.getOrganisationalUnit(`${organisationalUnit.id}/parent`).catch(error => {
-			if (error.response.status == 404) {
-				this.logger.debug(`Organisation ${organisationalUnit.id} has no parent`)
-			} else {
-				throw error
-			}
-		})
+	async getOrganisation(organisationalUnitId: number, includeParent: boolean = false): Promise<OrganisationalUnit> {
```

Comment:
May be worth having the second argument as an optional `options` object argument (eg. `{includeParents: true}` instead of a boolean?

Reply:
This was intentional - the `{includeParents: true}` parameters are the CSRS API interface parameters, it's useful for anything calling that (the organisationalUnitClient) to have access to those. However, anything calling this service method probably doesn't/shouldn't care about API params, so having a simple boolean here is more appropriate.

i.e if the API endpoint had 7 or so query string parameters added, we probably wouldn't want that to reflect into the service method, as it would add unnecessary complexity to anything calling it.

Reply:
Ok cool, no problem then. My only concern is when calling the `getOrganisation` function, you would call it like `getOrganisation(1, true)` [which could potentially be a code smell](https://martinfowler.com/bliki/FlagArgument.html).